### PR TITLE
Implement parsing for mask shorthand

### LINKED ISSUE #13235: Implement parsing for mask shorthand
I implemented all of the mask longhands in https://github.com/servo/servo/pull/13213. This means that we can parse things like `mask-image: url(foo)` or `mask-repeat: no-repeat`.

However, I deliberately left out the `mask` shorthand; so `mask: url(foo) no-repeat` doesn't work yet.

To make this work, we need to create a file `shorthands/mask.mako.rs` similar to https://github.com/servo/servo/blob/master/components/style/properties/shorthand/background.mako.rs

The mask and background shorthands are quite similar. They both encompass a set of array properties, of which there is a size, a position, an image, and a repeat. There are additional properties for both.

An important difference is that the mask shorthand doesn't have a concept akin to `background-color`. Whereas `mask` is a simple comma-separated list where each element is a bunch of longhands (where all mask longhands are allowed), `background` is a list of its component longhands where only the last element in the list is allowed to have a color. This should simiplify parsing a bit.

Spec: https://drafts.fxtf.org/css-masking-1/#the-mask

Grammar for single array element: `<mask-layer> = <mask-image> <mask-mode>? || <mask-position> [ / <mask-size> ]? ||<mask-repeat> || <mask-origin> || <mask-clip> || <mask-composite>`

The spec has more detailed steps on how to parse it, for example mask-origin with no defined mask-clip sets both.

You should be able to copy the background shorthand fine and start work from there, though.

This will need to be built with `./mach build-geckolib`, since mask is disabled for non-geckolib. I'll handle testing it on stylo. You can temporarily make the mask longhands not use `products="gecko"` to write unit tests.

### PR DESCRIPTION
**Pull request by canova**
Implement parsing for mask shorthand. It doesn't contain tests yet. I'll write and update the PR.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13235 (github issue number if applicable).

- [x] There are tests for these changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13336)

### CODE REVIEW

**Comment by Manishearth** [comment]
@canaltinova I found an easy way to disable the link failure.

I'm not sure if this is the right way to do it though, but it's good enough for now while you iterate, till I figure out something better.

```diff
diff --git a/tests/unit/stylo/fakebindings.rs b/tests/unit/stylo/fakebindings.rs
new file mode 100644
index 0000000..ef5e09d
--- /dev/null
+++ b/tests/unit/stylo/fakebindings.rs
@@ -0,0 +1,14 @@
+#[no_mangle]
+pub extern "C" fn Gecko_ReleaseURIArbitraryThread(aPtr: *mut ()) {
+    println!("Stub for Gecko_ReleaseURIArbitraryThread");
+}
+
+#[no_mangle]
+pub extern "C" fn Gecko_ReleasePrincipalArbitraryThread(aPtr: *mut ()) {
+    println!("Stub for Gecko_ReleasePrincipalArbitraryThread");
+}
+
+#[no_mangle]
+pub extern "C" fn Gecko_ReleaseAtom(aPtr: *mut ()) {
+    println!("Stub for Gecko_ReleaseAtom");
+}
\ No newline at end of file
diff --git a/tests/unit/stylo/lib.rs b/tests/unit/stylo/lib.rs
index a74d74d..8e75b0a 100644
--- a/tests/unit/stylo/lib.rs
+++ b/tests/unit/stylo/lib.rs
@@ -8,5 +8,6 @@ extern crate gecko_bindings;
 extern crate style;
 extern crate style_traits;
 
+pub mod fakebindings;
 mod sanity_checks;
 mod parsing;
```

**Comment by Manishearth** [comment]
@canaltinova also, looking at the tests, you don't need to test individual serialization like done in the style unit tests.

Just test that the `mask` shorthand parses to the right components. Perhaps also test that the mask serialization is correct; but you don't need to build mask from scratch the way tests/unit/style/properties/serialization does, just use the parsed values.

**Comment by Manishearth** [comment]
Another option is to continue using the style_tests crate for parsing tests, and compile it in a special mode where all properties are included. Don't touch the stylo tests crate then.

```diff
From 3ab46561a67dacc18c291012b9f696afb63b3dac Mon Sep 17 00:00:00 2001
From: Manish Goregaokar <manishsmail@gmail.com>
Date: Fri, 23 Sep 2016 09:40:16 +0530
Subject: [PATCH] Run tests with all properties enabled

---
 components/style/Cargo.toml          | 1 +
 components/style/build.rs            | 1 +
 components/style/properties/build.py | 4 +++-
 components/style/properties/data.py  | 7 ++++---
 4 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/components/style/Cargo.toml b/components/style/Cargo.toml
index 27dd885..14e412e 100644
--- a/components/style/Cargo.toml
+++ b/components/style/Cargo.toml
@@ -18,6 +18,7 @@ servo = ["serde", "serde/unstable", "serde_macros", "heapsize", "heapsize_plugin
          "cssparser/heap_size", "cssparser/serde-serialization",
          "selectors/heap_size", "selectors/unstable", "string_cache",
          "url/heap_size", "plugins"]
+testing = []
 
 [dependencies]
 app_units = "0.3"
diff --git a/components/style/build.rs b/components/style/build.rs
index 23ee9d5..ab385b8 100644
--- a/components/style/build.rs
+++ b/components/style/build.rs
@@ -55,6 +55,7 @@ fn main() {
         .arg(&script)
         .arg(product)
         .arg("style-crate")
+        .arg(if cfg!(feature = "testing") {"testing"} else {"regular"})
         .status()
         .unwrap();
     if !status.success() {
diff --git a/components/style/properties/build.py b/components/style/properties/build.py
index daa9140..6997588 100644
--- a/components/style/properties/build.py
+++ b/components/style/properties/build.py
@@ -23,10 +23,12 @@ def main():
         abort(usage)
     product = sys.argv[1]
     output = sys.argv[2]
+    testing = len(sys.argv) > 3 and sys.argv[3] == "testing"
+
     if product not in ["servo", "gecko"] or output not in ["style-crate", "geckolib", "html"]:
         abort(usage)
 
-    properties = data.PropertiesData(product=product)
+    properties = data.PropertiesData(product=product, testing=testing)
     rust = render(os.path.join(BASE, "properties.mako.rs"), product=product, data=properties)
     if output == "style-crate":
         write(os.environ["OUT_DIR"], "properties.rs", rust)
diff --git a/components/style/properties/data.py b/components/style/properties/data.py
index 94ae75f..b34a55b 100644
--- a/components/style/properties/data.py
+++ b/components/style/properties/data.py
@@ -160,8 +160,9 @@ class StyleStruct(object):
 
 
 class PropertiesData(object):
-    def __init__(self, product):
+    def __init__(self, product, testing):
         self.product = product
+        self.testing = testing
         self.style_structs = []
         self.current_style_struct = None
         self.longhands = []
@@ -179,7 +180,7 @@ class PropertiesData(object):
 
     def declare_longhand(self, name, products="gecko servo", **kwargs):
         products = products.split()
-        if self.product not in products:
+        if self.product not in products and not self.testing:
             return
 
         longhand = Longhand(self.current_style_struct, name, **kwargs)
@@ -194,7 +195,7 @@ class PropertiesData(object):
 
     def declare_shorthand(self, name, sub_properties, products="gecko servo", *args, **kwargs):
         products = products.split()
-        if self.product not in products:
+        if self.product not in products and not self.testing:
             return
 
         sub_properties = [self.longhands_by_name[s] for s in sub_properties]
-- 
2.8.3
```

 @emilio, any preferences? I like this latter option because we won't need to keep stubbing out methods.

**Comment by Manishearth** [comment]
Implemented the latter solution in https://github.com/servo/servo/pull/13386

You should rebase over that PR and just write your tests in the regular style unit tests crate.

**Comment by canaltinova** [comment]
Thanks for the help. I rebased your commits and added `use style::properties::shorthands::mask` to new mod inside serialization.rs, but when I try to run `./mach test-unit` I still get "no `mask` in `style::properties::shorthands`" error (Same for longhands). Am I missing something?

**Comment by Manishearth** [comment]
Got your ping about CSSStyleDeclaration -- you may need to temporarily edit CSSStyleDeclaration.webidl to add the mask shorthands and longhands.

However, you don't need to. `./mach build -p style` and `./mach test-unit -p style` should work without hitting this error. Since it's due to your temporarily enabling mask for both gecko and servo, we don't need to worry about the build failure.

**Comment by Manishearth** [comment]
https://github.com/servo/servo/pull/13415 should be a more permanent fix for your testing woes.

**Comment by Manishearth** [comment]
Re: your comment in IRC: `assert_roundtrip` won't work here, since that's just for longhands.

Instead, use the parse function you added, and assert that the `Longhands` struct formed has the correct values for the fields.

**Comment by Manishearth** [comment]
Code itself looks good, r=me on that, just needs some parse tests.

**Review by Manishearth — APPROVED** [review]
still needs the parse tests I mentioned, but otherwise okay

**Comment by canaltinova** [comment]
I added some tests about you've mentioned and fixed the nit. 
I have a doubt about origin and clip. Now, the clip is not setting in parsing function if just origin is set. Clip is setting the same as origin in serialization function. I choose this behavior because background is doing the same. I'm not sure it is the right behavior. What do you think?

**Comment by Manishearth** [comment]
Spec (https://drafts.fxtf.org/css-masking-1/#the-mask, https://drafts.csswg.org/css-backgrounds-3/#the-background) says that a single value should set both origin and clip.

The current code for both is buggy. However, that can be fixed later.

File: tests/unit/style/parsing/mask.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    let mut parser = Parser::new("padding-box padding-box");
+    let result = mask::parse_value(&context, &mut parser).unwrap();
+
+    assert_eq!(result.mask_origin.unwrap(), parse_longhand!(mask_origin, "padding-box"));
+    assert_eq!(result.mask_clip.unwrap(), parse_longhand!(mask_clip, "padding-box"));
+
+    let mut parser = Parser::new("padding-box");
+    let result = mask::parse_value(&context, &mut parser).unwrap();
+
+    assert_eq!(result.mask_origin.unwrap(), parse_longhand!(mask_origin, "padding-box"));
```
Comment (Manishearth) [inline tests/unit/style/parsing/mask.rs]: File a followup about the clip/origin issue and leave a comment here linking to that.

**Comment by Manishearth** [comment]
r=me with the bug filed and links

**Comment by canaltinova** [comment]
filed an issue here: #13466 and added comment about it.

**Comment by Manishearth** [comment]
Oh, #13415 blocks this

**Comment by canaltinova** [comment]
Fixed merge conflicts.

**Comment by Manishearth** [comment]
wooo!

**Comment by canaltinova** [comment]
🎉 🎉 🎉

**Pull request merged**